### PR TITLE
More async workflows

### PR DESCRIPTION
**Pull request by conte91**
Add asynchronous password entry and unlock workflows. Depends on #312.

### CODE REVIEW

File: src/workflow/unlock.h [hunk 1]
```diff
@@ -15,15 +15,31 @@
...
 
+#include "workflow.h"
+
 #include <compiler_util.h>
 #include <keystore.h>
 
+/**
+ * Create a workflow that prompts the user for the password and unlocks the keystore, calls
+ * the provided callback, and exits. The workflow will terminate if either the user enters the
+ * correct password, or the device is reset after too many failed attempts. The callback will
```
Comment (benma) [inline src/workflow/unlock.h]: > or the device is reset after too many failed attempts.

In this case the device reboots and there won't be any response, right?

What about if HWW_REQ_CANCEL comes in?

File: src/util.c [hunk 2]
```diff
@@ -58,6 +59,17 @@ void util_cleanup_64(uint8_t** buf)
     util_zero(*buf, 64);
 }
 
+char* util_strdup(const char* str)
+{
+    size_t str_size = strlen(str);
+    char* result = malloc(str_size + 1);
```
Comment (benma) [inline src/util.c]: why not use `strdup` here?

File: src/workflow/unlock.c [hunk 3]
```diff
@@ -122,37 +143,143 @@ keystore_error_t workflow_unlock_and_handle_error(const char* password)
...
+        util_zero(data->password, sizeof(data->password));
+        free(data->password);
+        data->password = NULL;
+    }
+}
+
+/**
+ * Callback invoked when the password entry completes.
+ */
+static void _unlock_password_completed(const char* password, void* param)
```
Comment (benma) [inline src/workflow/unlock.c]: not about this particular callback, but in general: before i went through a lot of effort to reduce "callback hell", and now it seems we are going callback heavy again.

it might be required, but it would be nice if we could figure out something that is easier to read/follow than lots of ad-hoc state machines and callbacks. 

before, the callbacks were hard to maintain and bug-prone to a point where we did not introduce new workflows because of it.

cc @NickeZ

File: src/workflow/unlock.h (hunk 1 above)
Comment (conte91) [reply src/workflow/unlock.h]: If the workflow is stopped, cleanup() will be called. No further action needed (as every active workflow cleans up after itself). Note that atm there is no `abort()` mechanism, so stopping a workflow from "outside" is still out of scope for this PR. That will come with #164. Forcefully aborting a workflow will just destroy (`workflow_stop_workflow()`) every workflow in the stack until the selected one, which should guarantee a proper cleanup no matter the state.

So tl;dr: 
> What about if HWW_REQ_CANCEL comes in?

This is not responsibility of the `workflow` stack :)

File: src/util.c (hunk 2 above)
Comment (conte91) [reply src/util.c]: Something something because portability (nah, it's because I'm stupid. Will fix).

File: src/workflow/unlock.c (hunk 3 above)
Comment (conte91) [reply src/workflow/unlock.c]: Discussion ongoing offline

File: src/util.c (hunk 2 above)
Comment (conte91) [reply src/util.c]: Fixed! 3278d241

File: test/unit-test/framework/workflows/workflow_status.c [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+void _mock_status_spin(workflow_t* self)
+{
+    data_t* data = (data_t*)self->data;
+    if (data->callback) {
+        data->callback(data->callback_param);
+    }
+    workflow_stack_stop_workflow();
+}
+
+workflow_t* mock_workflow_status(void (*callback)(void*), void* callback_param)
```
Comment (NickeZ) [inline test/unit-test/framework/workflows/workflow_status.c]: why can't you use the real `workflow_status`?

File: src/workflow/unlock_bip39.c [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+{
+    unlock_bip39_data_t* data = (unlock_bip39_data_t*)self->data;
+    { // animation
+        // Cannot render screens during unlocking (unlocking blocks)
+        // Therefore hardcode a status screen
+        UG_ClearBuffer();
+        image_lock(SCREEN_WIDTH / 2, SCREEN_HEIGHT / 2 - 1, IMAGE_DEFAULT_LOCK_RADIUS);
+        UG_SendBuffer();
+#ifndef TESTING
+        delay_ms(1200);
```
Comment (NickeZ) [inline src/workflow/unlock_bip39.c]: what the heck, do we artificially increase the processing time here? This is bad for u2fhid compatability no?

File: src/workflow/unlock.c (hunk 3 above)
Comment (NickeZ) [reply src/workflow/unlock.c]: As discussed in other places, now that we have proper asynchronous workflows we can't have code that look synchronous/sequential since C doesn't have support for async.

I'm totally fine with this PR.

File: src/workflow/unlock_bip39.c (hunk 5 above)
Comment (conte91) [reply src/workflow/unlock_bip39.c]: Yep, it is, and should be easy  to fix. I guess it's just to have a "nice" animation.

File: test/unit-test/framework/workflows/workflow_status.c (hunk 4 above)
Comment (conte91) [reply test/unit-test/framework/workflows/workflow_status.c]: This is used by the `workflow_cancel` unit test, so the status workflow is mocked (it already was, but had to mock the async version as well!) to make it a "proper" _unit_ test and avoid testing stuff that doesn't belong to it.

File: src/workflow/get_mnemonic_passphrase.c [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+{
+    char* passphrase_out = (char*)param;
+    int n_written = snprintf(passphrase_out, SET_PASSWORD_MAX_PASSWORD_LENGTH, "%s", passphrase);
+    if (n_written < 0 || n_written >= SET_PASSWORD_MAX_PASSWORD_LENGTH) {
+        Abort("Mnemonic passphrase invalid length.");
+    }
+    workflow_blocking_unblock();
+}
+
+void get_mnemonic_passphrase_blocking(char* passphrase_out)
```
Comment (benma) [inline src/workflow/get_mnemonic_passphrase.c]: unused

File: src/workflow/get_mnemonic_passphrase.c [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+/**
+ * Invoked when the user has finished with the visual
+ * passphrase confirmation.
+ */
+static void _confirm_passphrase_cb(bool result, void* param)
+{
+    mnemonic_passphrase_data_t* data = (mnemonic_passphrase_data_t*)param;
+    if (!result) {
+        /* Confirmation failed. Start over. */
+        _clean_passphrase(data);
```
Comment (benma) [inline src/workflow/get_mnemonic_passphrase.c]: nit: a little bit brittle, would not be needed if the passphrase was not on the heap but fixed size allocated with `SET_PASSWORD_MAX_PASSWORD_LENGTH`

File: src/workflow/unlock.c [hunk 8]
```diff
@@ -118,37 +73,129 @@ keystore_error_t workflow_unlock_and_handle_error(const char* password)
...
     if (!keystore_is_locked()) {
-        return true;
+        data->state = UNLOCK_STATUS_FINISHED;
+        data->result = true;
+        return;
     }
 
+    /*
+     * FUTURE: maybe remove?
+     * It might be a "security" feature/hack, but it's not
```
Comment (benma) [inline src/workflow/unlock.c]: it's not a security feature, it's a "failsafe": 

- some screen is launched and on the stack (e.g. the random number call)
- app quits
- app restarts
- app sends unlock command
- old screen is still there, op_unlock pushes on top of it, and after it's popped, the old screen reappears

pop_all() is/was here because a new app/wallet will call op_unlock as one of the first things.

Not sure if we can get rid of it yet.

File: src/workflow/unlock.c [hunk 9]
```diff
@@ -13,86 +13,37 @@
...
-#include <ui/components/ui_images.h>
-#include <ui/fonts/password_11X12.h>
 #include <ui/screen_stack.h>
-#include <ui/ugui/ugui.h>
+#include <ui/workflow_stack.h>
 #include <util.h>
-#ifndef TESTING
-#include <hal_delay.h>
-#endif
+#include <workflow/get_mnemonic_passphrase.h>
```
Comment (benma) [inline src/workflow/unlock.c]: does not seem used in this file?

File: src/workflow/get_mnemonic_passphrase.h [hunk 10]
```diff
@@ -0,0 +1,11 @@
+#ifndef _GET_MNEMONIC_PASSPHRASE_H
+#define _GET_MNEMONIC_PASSPHRASE_H
+
+#include <stdbool.h>
+
+#include "workflow.h"
+
+workflow_t* get_mnemonic_passphrase(void (*callback)(char*, void*), void* callback_param);
```
Comment (benma) [inline src/workflow/get_mnemonic_passphrase.h]: should be `workflow_` prefixed

File: src/workflow/unlock.c (hunk 8 above)
Comment (conte91) [reply src/workflow/unlock.c]: That can't happen with the current code, can it?
* Blocking workflow (e.g. random number) starts
* App quits
* App restarts
* App sends unlock commands -> USB stack is blocked, command is dropped

Or with the async workflows:
* Async workflow (e.g. random number) starts, locks the USB stack
* App quits
* App restarts
* App sends unlock commands -> USB stack is blocked, command is rejected

And after implementing timers:
* Async workflow (e.g. random number) starts, locks the USB stack
* App quits
* App restarts
* App sends unlock commands -> USB stack is blocked, command is rejected
* Original workflow times out after a bit -> USB stack unlocked
* App retries the unlock command -> Unlock workflow starts.

Comment (NickeZ) [reply src/workflow/unlock.c]: > pop_all() is/was here because a new app/wallet will call op_unlock as one of the first things.
> 
> Not sure if we can get rid of it yet.

It would be great if we never have to do `pop_all()`.

File: src/workflow/get_mnemonic_passphrase.c (hunk 6 above)
Comment (conte91) [reply src/workflow/get_mnemonic_passphrase.c]: Removed 9d82e082

File: src/workflow/unlock.c (hunk 9 above)
Comment (conte91) [reply src/workflow/unlock.c]: 133648c0

File: src/workflow/get_mnemonic_passphrase.h (hunk 10 above)
Comment (conte91) [reply src/workflow/get_mnemonic_passphrase.h]: 13dac1e2
